java/src/main/java/com/google/protobuf/ZeroCopyEncoder.java

```diff
@@ -0,0 +1,647 @@
+// Protocol Buffers - Google's data interchange format
+// Copyright 2015 Google Inc.  All rights reserved.
+// https://developers.google.com/protocol-buffers/
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.protobuf;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A protobuf {@link Encoder} that does little to no buffering for fields. It's not fully
+ * zero-copy in that it does maintain an internal buffer for small operations. Large fields,
+ * however, are written directly through.
+ */
+public final class ZeroCopyEncoder implements Encoder {
+  // TODO(nmittler): Consider using an allocator for buffers to allow pooling.
+  // TODO(nmittler): Need experimentation with good values.
+  private static final int SMALL_BUFFER = 1024;
+  private static final double COPY_THRESHOLD = 0.75;
+
+  /**
+   * Java 1.8 defines Integer.BYTES, but we need this for older versions.
+   */
+  private static final int INTEGER_BYTES = Integer.SIZE / 8;
+  private static final int LONG_BYTES = Long.SIZE / 8;
+
+  private final Handler handler;
+  private final int bufferSize;
+
+  /**
+   * Internal buffer used to avoid writing small chunks of data. This is guaranteed to be
+   * backed by an array.
+   */
+  private ByteBuffer buffer;
```

Comment:
We currently have to use a heap buffer since we need an array for some of the `Utf8` and `ByteString` methods.  We can look into adding alternative methods that support `ByteBuffer` to allow us to leverage direct buffers.  However, I would expect that work to be done as follow-on work, especially since it won't impact the interface.